rspace/src/main/scala/coop/rchain/rspace/history/TrieCache.scala

```diff
@@ -0,0 +1,181 @@
+package coop.rchain.rspace.history
+import coop.rchain.rspace.Blake2b256Hash
+
+import scala.collection.immutable.{Map => ImmutableMap}
+import scala.collection.immutable.Seq
+import scala.collection.mutable.Map
+
+private[rspace] trait CachedItem[TItem] {
+  def asOption: Option[TItem]
+}
+
+private[rspace] case class LoadedItem[TItem](value: TItem) extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = Some(value)
+}
+
+private[rspace] case class StoredItem[TItem](value: TItem) extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = Some(value)
+}
+
+private[rspace] case class AbsentItem[TItem]() extends CachedItem[TItem] {
+  override def asOption: Option[TItem] = None
+}
+
+private[rspace] class TrieCache[T, K, V](
+    trieStore: ITrieStore[T, K, V],
+    val trieBranch: Branch,
+    trieStoreOwner: Boolean = false
```

Comment:
what's the purpose of this flag? (I did not see it used other then the default)

Reply:
If it's enabled closing the cache closes also the underlying store. This might be convenient if we decide to create TrieCache-specific unit-tests or storages.